### PR TITLE
Update Experimental Configuration automatically

### PR DESCRIPTION
**Pull request by rhomspuron**
See #806

### CODE REVIEW

**Comment by rhomspuron** [comment]
Hi @dschick, thanks for test it!!

>  would it be possible force overwriting the expconfig without the popup?

Yes, it's possible. But we decided not to do it, this is the policy used in general. The point is, if the user is not aware of who or what generates the change (a macro or another user in another door), it will be a problem identify if it was changed, so we show the popup. 

The same policy is implemented on the TaurusValue write widget, if the write value is different on the server and the client, the widget change the font color to blue.

spectacle q16414


> Only the Trend Window does not updated the plotted channels, but I guess this is another issue?

Yes, this is another issue. We need to connect the change events to the plot too.

**Comment by dschick** [comment]
hej @rhomspuron 

okay, I guess that is a wise policy. I think that most people are either working with the gui or with the cli in order to change the expconf and in case one is mixing stuff than the pop-up will be a good reminder.

Do you think you can handle the change event for the plot too?
For me, this is a quite important issue for usability - either that changing the plotted counters work directly from the expconf or from the cli via the `plotselect `macro #824

**Comment by rhomspuron** [comment]
>  okay, I guess that is a wise policy. I think that most people are either working with the gui or with the cli in order to change the expconf and in case one is mixing stuff than the pop-up will be a good reminder.

About this, @reszelaz commented me one idea. We can include on the expconf launcher an option like --autofix or --automatic-update, and force the update when we receive the event. What do you think?

**Comment by rhomspuron** [comment]
>  Do you think you can handle the change event for the plot too?

Yes, we will do it ASAP.

**Comment by dschick** [comment]
@rhomspuron the --auto-update parameter would be great idea, especially in a lab environment where server and client normally run on the same machine.

**Comment by dschick** [comment]
so I update from pyTango 9.2.2 to 9.2.4 and taurus was 4.4.0
Tango comes from the ubuntu repos as I using MINT 19

Could you try upgrading to pytango 9.2.4 because I think that it was working before I upgraded.

Thanks

Daniel

**Comment by rhomspuron** [comment]
Did you restart the spock too?

**Comment by dschick** [comment]
no, I did not restart spock, but I noticed that now that the window was always popping up after a change of the config AND starting a scan - but not directly after changing the config.

Moreover the expconf and trend window were quite unresponsive today and the message window asking for keeping or loading changes into the expconfig always hung completely. However, I do not know if this has something todo with my setup and PC but generally I did not encounter something like that before I merged this PR. I will try tomorrow again after a restart of the machine.
I can also try downgrading pytango

**Comment by rhomspuron** [comment]
>   I noticed that now that the window was always popping up after a change of the config AND starting a scan

Yes, the pop-up will appear by any changes on the common environment variables and the configuration of the measurement groups. After your recommendation we implemented the --auto-update, could you test it?  The way is: ```%expconf --auto-update ```

>  but not directly after changing the config

I noticed that some times in my PC the event takes a time, but I correlated it with network problems. 

>  the message window asking for keeping or loading changes into the expconfig always hung completely

I will test it again, but the pop-up did not hung the expconf in my case.

**Comment by dschick** [comment]
after a restart of the machine it runs much smoother today.
however:

> Yes, the pop-up will appear by any changes on the common environment variables and the configuration of the measurement groups.

This is not true for me. As I wrote when I execute the `plotselect` macro #824 I do change the env but the pop-up sometimes appears after every `plotselect` and sometimes it does not at all.

So after I executed the `plotselect` like 10 time the expconf and trend windows get quite unresponsive which might also be the reason why I do not get the pop-up

**Comment by dschick** [comment]
> the way is: %expconf --auto-update

this works as expected but with the same problems as without the --auto-update as described above

**Comment by dschick** [comment]
see my comments in #917

**Comment by rhomspuron** [comment]
Hi @dschick,

I just merger develop to test the PR #897. I will try to reproduce the problem, but maybe it is a problem of the expconf plot.

**Comment by dschick** [comment]
yeah, it seems pretty much, that its about the expconf plot, however, when I test it on sar_demo door, the expconf plot works, when this PR is NOT merged.
In any other case, the expconf plot freezes with or w/o this PR merged.

BTW, right now, I do not get any pop-ups when externally changing the expconf. I think this behaviour exists since the last release.

**Comment by rhomspuron** [comment]
>  BTW, right now, I do not get any pop-ups when externally changing the expconf. I think this behaviour exists since the last release.

I checked it and it does not work more, I will investigate what produce the problem, maybe the #897

**Comment by rhomspuron** [comment]
Hi @dschick 

We found the bug, the problem was multiple connections to the same Qt signals. I ran 400 scans without problems. Could you test it?

**Comment by dschick** [comment]
@rhomspuron 

> It started to work again after to start from the scratch. I removed the Pool and MS instance from the TangoDB and deleted the MS environment (/tmp/tango/xxx). I suppose that I had the problem on my setup like @reszelaz commented on #897 

hm, this would be a quite nasty if I would need to delete and create all of my configuration again from scratch. Is there a way how to dump Pool and Macroserver in Tango - I asked a similar question already in #836 ?
Of course the dump might then also contain the faulty configuration ... okay, so it seems I have to do the nasty job

**Comment by dschick** [comment]
hm, I just created a new door for testing with sar_demo and it is still not working when I use the `plotselect` macro. The pop-up came up once when closing the sardana terminal but I cannot reproduce that.

**Comment by rhomspuron** [comment]
I will test your macro plotselect and try to reproduce the problem.

File: src/sardana/taurus/qt/qtcore/tango/sardana/macroserver.py [hunk 1]
```diff
@@ -103,8 +107,53 @@ def logReceived(self, log_name, output):
...
+
 
 class QMacroServer(BaseMacroServer, Qt.QObject):
+    # TODO: For Taurus 4 compatibility
+    try:
+        typesUpdated = Qt.pyqtSignal()
+        elementsUpdated = Qt.pyqtSignal()
+        elementsChanged = Qt.pyqtSignal()
+        macrosUpdated = Qt.pyqtSignal()
+        environmentChange = Qt.pyqtSignal(list)
```
[the remark this answers is not in the document]
Comment (rhomspuron) [reply src/sardana/taurus/qt/qtcore/tango/sardana/macroserver.py]: Ok

File: src/sardana/taurus/qt/qtgui/extra_sardana/expdescription.py [hunk 2]
```diff
@@ -173,6 +239,117 @@ def __init__(self, parent=None, door=None, plotsButton=True):
...
+        btn_ok.setText('Load')
+        btn_cancel = self._expConfChangedDialog.button(Qt.QMessageBox.Cancel)
+        btn_cancel.setText('Keep')
+        result = self._expConfChangedDialog.exec_()
+        self._expConfChangedDialog = None
+        if result == Qt.QMessageBox.Ok:
+            self._reloadConf(force=True)
+
+    @QtCore.pyqtSlot()
+    def _experimentalConfigurationChanged(self):
```
[the remark this answers is not in the document]
Comment (rhomspuron) [reply src/sardana/taurus/qt/qtgui/extra_sardana/expdescription.py]: I'll change it

**Comment by rhomspuron** [comment]
Thanks for the review @reszelaz, I will check the creation of the measurement group.

**Comment by reszelaz** [comment]
With the recent changes everything seems to work as expected. Merging it..

**Comment by dschick** [comment]
works for me as well
many thanks!

**Pull request merged**